RFE: Sync to Linux 6.12 syscall definitions

xen0n opened the pull request:
The `arch-syscall-validate` script updates will need to be backported to v2.5.x later, for continuing to sync aarch64 and riscv64 that got refactored to source the syscall definitions from a `syscall.tbl` instead of `asm-generic/unistd.h`.

xen0n commented:
After the change, `linux-headers >= 6.11` will be required for native LoongArch builds, for the `fstat` `newfstatat` definitions to match (for test 38 and 55 to pass).

lixing-star commented:
> After the change, `linux-headers >= 6.11` will be required for ~native LoongArch~ all builds, for definitions of new syscalls introduced in this version.

this is really helpful for LoongArch, thanks.

Dandan336 commented:
Hi maintainers,

- About defining __ARCH_WANT_NEW_STAT in unistd.h
'LoongArch: Define __ARCH_WANT_NEW_STAT in unistd.h' [1] **was merged in linux upstream** [2].
'LoongArch: Define __ARCH_WANT_NEW_STAT in unistd.h' **was supported in the stable version of linux**, as follows,

1. branch linux-6.1.y
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.1.y&id=68a35d0abf8059ee41f164093c700297343101b2
2. branch linux-6.6.y
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.6.y&id=9d91b004df9a6aa2487661e52a398fd808184955
3. branch linux-6.10.y
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.10.y&id=268a625399c63083aeb9c4fc818f5f2d916d5adb
4. branch linux-6.11.y
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-6.11.y&id=7697a0fe0154468f5df35c23ebd7aa48994c2cdc

[1]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7697a0fe0154468f5df35c23ebd7aa48994c2cdc
[2]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=a362ade892e3e4de69296cddb1a23a1efe701428

- The results of the test cases 
Based on main branch of libseccomp, I pulled this pull-request 'RFE: Sync to Linux 6.11 syscall definitions' submitted by xen0n. I built libseccomp project on Debian loong64 rootfs (the verson of linux-libc-dev is 6.11.2-1), **there is 0 failed or errored test case**.The results of the test cases are as follows,
```
Regression Test Summary
 tests run: 5084
 tests skipped: 124
 tests passed: 5084
 tests failed: 0
 tests errored: 0
=================================
PASS: regression
=============
1 test passed
=============
```
Please review.

yetist commented:
@drakenclimber  @pcmoore 

Could you review this PR? Thanks.
Also, I would like to know when v2.6.0 will be released.

rusty-snake commented:
> Also, I would like to know when v2.6.0 will be released.

https://github.com/seccomp/libseccomp?tab=readme-ov-file#release-process

TL;DR: When it's ready.

pcmoore commented on include/seccomp-syscalls.h:
```diff
@@ -1097,6 +1118,12 @@
+#else
+#define __SNR_map_shadow_stack		__PNR_map_shadow_stack
+#endif
+
```
Isn't `map_shadow_stack()` defined for all the arches?

xen0n replied:
Yes it is, but in [this commit](https://github.com/seccomp/libseccomp/commit/53267af3fb56eed93a50b8ef92f41825c97a7813) the PNR was allocated anyway and the change is already shipped in v2.5.5. So if we're to consider PNR values as part of the public API we would have to keep `__PNR_map_shadow_stack` defined, and for stylistic consistency we would have to keep this `#ifdef`-fery as well.

pcmoore replied:
Thanks for the reminder @xen0n.

In the manpages we "highly recommend" the use of the `SCMP_SYS()` macro as opposed to directly specifying syscall numbers so it's a bit of a gray area.  I think it would be okay for us to modify/remove the PNR syscall values if necessary, but we probably shouldn't do so unless we have a good reason.  Considering everything, it looks like you approach of leaving the PNR values as-is is likely the right one.  Thanks :)

pcmoore commented:
Just one small comment/clarification/question above, but otherwise I think this looks good.  Thanks for all your work on this @xen0n!

rusty-snake commented on include/seccomp-syscalls.h:
```diff
@@ -280,6 +280,9 @@
 #define __PNR_atomic_barrier			-10246
 #define __PNR_atomic_cmpxchg_32			-10247
 #define __PNR_getpagesize			-10248
+#define __PNR_map_shadow_stack			-10249
```
FTR: The PNR numbers below `-10244` (__PNR_memfd_secret) have already diverged between [`libseccomp:main`](https://github.com/seccomp/libseccomp/blob/main/include/seccomp-syscalls.h#L278) and [`libseccomp:release-2.5`](https://github.com/seccomp/libseccomp/blob/release-2.5/include/seccomp-syscalls.h#L278).

pcmoore replied:
As a FYI, I'm okay with fixing that sort of thing while merging the PR.

xen0n replied:
I have preserved the `__PNR_map_shadow_stack` position while rebasing.

xen0n commented:
This is now rebased and updated to reflect Linux v6.12-rc5 syscall info (no new syscalls since 6.11). I have also incorporated @rusty-snake's review comment of keeping PNR allocations consistent with v2.5.x. Tests still pass on aarch64, loongarch64, riscv64 and x86\_64.

drakenclimber commented:
My apologies for being quiet the last few weeks.  I've had some other high-priority work on my plate, but I'm thinking I should have some time in the next few weeks to get back into libseccomp work.  Thanks all for your patience :).

pcmoore commented:
This looked good to me and it passed all my tests locally so I merged the PR at HEAD f01e67509e45c672f4bdd643d94d90867cc19d90, thanks for you work and patience @xen0n!

@drakenclimber nothing in here looked controversial and we needed something like this for future syscall updates so I went ahead and merged the PR, if you notice anything awry please let me know and we can revert/fix it.

pcmoore commented:
While the PR is merged, I'm going to leave this open until I finish the backport to the release-2.5 branch.

pcmoore commented:
Here is the release-2.5/v2.5.x branch backport: https://github.com/seccomp/libseccomp/pull/439
